numba_dpex/core/utils/reduction_helper.py

```diff
@@ -0,0 +1,446 @@
+                and isinstance(param_types[pindex], types.npytypes.Array)
+            ):
+                parfor_params.append(parfor_params_orig[pindex] + "param")
+            else:
+                parfor_params.append(parfor_params_orig[pindex])
+
+        # Change parfor body to replace illegal loop index vars with legal ones.
+        replace_var_names(loop_body, ind_dict)
+        self.legal_loop_indices = legal_loop_indices
+        self.loop_body = loop_body
+        self.func_arg_types = func_arg_types
+        self.ind_dict = ind_dict
+        self.param_dict = param_dict
+        self.parfor_legalized_params = parfor_params
+        self.param_types = param_types
+        self.lowerer = lowerer
+        self.work_group_size = reductionHelperList[0].work_group_size
+
+    def copy_final_sum_to_host(self, psrfor_kernel):
+        lowerer = self.lowerer
+        ir_builder = KernelLaunchIRBuilder(lowerer, psrfor_kernel.kernel)
+
+        # Create a local variable storing a pointer to a DPCTLSyclQueueRef pointer.
+        curr_queue = ir_builder.get_queue(exec_queue=psrfor_kernel.queue)
+
+        builder = lowerer.builder
+        context = lowerer.context
+
+        memcpy_fn = DpctlCAPIFnBuilder.get_dpctl_queue_memcpy(
+            builder=builder, context=context
+        )
+        event_del_fn = DpctlCAPIFnBuilder.get_dpctl_event_delete(
+            builder=builder, context=context
+        )
+        event_wait_fn = DpctlCAPIFnBuilder.get_dpctl_event_wait(
+            builder=builder, context=context
+        )
+
+        for i, redvar in enumerate(self.parfor_redvars):
+            srcVar = self.final_sum_names[i]
+
+            item_size = builder.gep(
+                lowerer.getvar(srcVar),
+                [
+                    context.get_constant(types.int32, 0),
+                    context.get_constant(types.int32, 3),  # itmesize
+                ],
+            )
+
+            array_attr = builder.gep(
+                lowerer.getvar(srcVar),
+                [
+                    context.get_constant(types.int32, 0),
+                    context.get_constant(types.int32, 4),  # data
+                ],
+            )
+
+            dest = builder.bitcast(
+                lowerer.getvar(redvar),
+                utils.get_llvm_type(context=context, type=types.voidptr),
+            )
+            src = builder.bitcast(
+                builder.load(array_attr),
+                utils.get_llvm_type(context=context, type=types.voidptr),
+            )
+
+            args = [
+                builder.load(curr_queue),
+                dest,
+                src,
+                builder.load(item_size),
+            ]
+
+            event_ref = builder.call(memcpy_fn, args)
+
+            builder.call(event_wait_fn, [event_ref])
+            builder.call(event_del_fn, [event_ref])
+
+        ir_builder.free_queue(sycl_queue_val=curr_queue)
+
+
+def update_sentinel(gufunc_ir, sentinel_name, loop_body, new_label):
```

Comment:
This function is used by all types of parfors and not only reductions, can be moved into the `kernel_builder` module